### PR TITLE
Support for FreeMiNT on Atari computers and clones/VMs

### PR DESCRIPTION
**Pull request by ArmstrongJ**
## Description

This pull request adds support for FreeMiNT, a UNIX-like operating system for Atari ST/TT/Falcon computers and related clone machines.  FreeMiNT normally uses GEM for its graphical user interface, so the detection routines for its window manager maps to the AES, or Application Environment Services, rather than an X window manager.

Basically, all features are properly supported. Packages are assumed to be based on RPMs for this OS.  The terminal, TosWin2, used in almost all cases on FreeMiNT, is not ideal, so there are a few additions and omissions to get things looking acceptable on it.  

The ASCII graphic for the distro is the GEM "busy bee," which avoids using the Atari logo.  It is commonly used in the Atari community, especially on the current [Firebee Atari clone](http://firebee.org/fb-bin/index).

I'll be happy to test or add any features necessary.  I've tested this under the ARAnyM virtual machine.  I've attached a sample screen capture.

neofetch-freemint

### CODE REVIEW

File: neofetch [hunk 1]
```diff
@@ -426,6 +435,13 @@ get_uptime() {
                     h="${h#0}" t="${t#0}"
                     seconds="$((d*86400 + h*3600 + ${t%%:*}*60 + ${t#*:}))"
                 ;;
+
+                "FreeMiNT")
+                    boot="$(sysctl -n kern.boottime)"
+                    now="$(date +%s)"
+                    seconds="$((now - boot))"
+                ;;
+
```
Comment (dylanaraps) [inline neofetch]: Remove this line.

File: neofetch [hunk 2]
```diff
@@ -604,6 +620,12 @@ get_packages() {
         "IRIX")
             packages="$(($(versions -b | wc -l)-3))"
         ;;
+
+        "FreeMiNT")
+            type -p rpm >/dev/null && \
+                packages="$((packages+=$(rpm -qa | wc -l)))"
+        ;;
+
```
Comment (dylanaraps) [inline neofetch]: Remove this line

File: neofetch [hunk 3]
```diff
@@ -653,6 +675,26 @@ get_de() {
             esac
         ;;
 
+        "FreeMiNT")
+            get_wm
+
+            if [ -e /proc/thing* ]; then
```
Comment (dylanaraps) [inline neofetch]: Double `[[` `]]` and quote the dir name.

File: neofetch [hunk 4]
```diff
@@ -1104,6 +1161,12 @@ get_cpu() {
             # Get CPU cores.
             cores="$(sysconf NPROC_ONLN)"
         ;;
+
+        "FreeMiNT")
+            cpu="$(cat /kern/cpuinfo | awk -F':' '/CPU:/ {printf $2}')"
```
Comment (dylanaraps) [inline neofetch]:
```
cpu="$(awk -F':' '/CPU:/ {printf $2}' /kern/cpuinfo)"
```

File: neofetch [hunk 5]
```diff
@@ -1104,6 +1161,12 @@ get_cpu() {
...
+            speed="$(cat /kern/cpuinfo | awk -F':' '/Clocking:/ {printf $2}' | awk -F'.' '{ printf $1 }')"
```
Comment (dylanaraps) [inline neofetch]:
```
+            speed="$(awk -F':' '/Clocking:/ {printf $2}' | awk -F'.' '{ printf $1 }' /kern/cpuinfo)"

```

Comment (dylanaraps) [reply neofetch]: Can this not be reduced to one `awk` statement?

File: neofetch [hunk 6]
```diff
@@ -1445,6 +1509,15 @@ get_memory() {
             mem_free="$((mem_stat[5] / 1024))"
             mem_used="$((mem_total - mem_free))"
         ;;
+
+        "FreeMiNT")
+            mem_total="$(cat /kern/meminfo | awk -F ':|kB' '/MemTotal:/ {printf $2}')"
```
Comment (dylanaraps) [inline neofetch]:
```
mem_total="$(awk -F ':|kB' '/MemTotal:/ {printf $2}' /kern/meminfo)"
```

File: neofetch [hunk 7]
```diff
@@ -1445,6 +1509,15 @@ get_memory() {
...
+            mem_free="$(cat /kern/meminfo | awk -F ':|kB' '/MemFree:/ {printf $2}')"
```
Comment (dylanaraps) [inline neofetch]:
```
mem_free="$(awk -F ':|kB' '/MemFree:/ {printf $2}' /kern/meminfo)"
```

File: neofetch [hunk 8]
```diff
@@ -1445,6 +1509,15 @@ get_memory() {
...
+            mem_used="$(( mem_total - mem_free ))"
```
Comment (dylanaraps) [inline neofetch]: Remove the spaces around the `$(())`

File: neofetch [hunk 9]
```diff
@@ -1833,6 +1906,11 @@ get_term() {
         *) term="${TERM_PROGRAM/\.app}" ;;
     esac
 
+    # Most likely TosWin2 on FreeMiNT - quick check
+    if [[ "$TERM" == "tw52" || "$TERM" == "tw100" ]]; then
```
Comment (dylanaraps) [inline neofetch]:
```
[[ "$TERM" == "tw52" || "$TERM" == "tw100" ]] && \
    term="TosWin2"
```

File: neofetch [hunk 10]
```diff
@@ -2384,6 +2462,12 @@ get_cols() {
 
     unset -v blocks blocks2 cols
 
+    # TosWin2 on FreeMiNT is terrible at this,
+    # so we'll reset colors arbitrarily
+    if [ "${term}" == "TosWin2" ]; then
```
Comment (dylanaraps) [inline neofetch]:
```
[[ "$term" == "TosWin2" ]] && \
    printf "%b\n" "\\e[30;47m"
```

File: neofetch [hunk 11]
```diff
@@ -763,6 +805,21 @@ get_wm() {
                 [[ "$wm" == "blackbox" ]] && wm="bbLean (Blackbox)"
                 wm="${wm:+$wm, }Explorer"
             ;;
+
+            "FreeMiNT")
+                if [ -e /proc/xaaes* ]; then
```
Comment (dylanaraps) [inline neofetch]: Double [[ ]] and quote the dir name.

File: ascii/distro/gem [hunk 12]
```diff
@@ -0,0 +1,19 @@
+
```
Comment (dylanaraps) [inline ascii/distro/gem]: Remove all padding around the art. The art should touch the left and top edges of the file.

File: neofetch [hunk 13]
```diff
@@ -1511,11 +1500,11 @@ get_memory() {
         ;;
 
         "FreeMiNT")
-            mem_total="$(cat /kern/meminfo | awk -F ':|kB' '/MemTotal:/ {printf $2}')"
-            mem_free="$(cat /kern/meminfo | awk -F ':|kB' '/MemFree:/ {printf $2}')"
-            mem_used="$(( mem_total - mem_free ))"
-            mem_total="$(( mem_total / 1024 ))"
-            mem_used="$(( mem_used / 1024 ))"
+            mem_total="$(awk -F ':|kB' '/MemTotal:/ {printf $2}' /kern/meminfo)"
```
Comment (dylanaraps) [inline neofetch]: These two `awk | /kern/meminfo` can probably become one statement.

```sh
mem="$(awk -F ':|kB' '/MemTotal:|MemFree:/ {printf $2, " "}' /kern/meminfo)"
mem_used="${mem/* }"
mem_total="${mem/ *}"
```

File: neofetch [hunk 14]
```diff
@@ -807,19 +800,16 @@ get_wm() {
...
-                    wm="MyAES"
-                elif [ -e /proc/naes* ]; then
-                    wm="N.AES"
-                elif [ -e /proc/geneva* ]; then
-                    wm="Geneva"
-                else
-                    wm="Atari AES"
-                fi
+                wm="Atari AES"
+                for possibles in "$(ls /proc | awk -F'[ .]' '/aes|geneva/ {print $1}')"; do
```
Comment (dylanaraps) [inline neofetch]:
```
for files in /proc/*; do
```

File: neofetch [hunk 15]
```diff
@@ -678,21 +676,16 @@ get_de() {
...
-            elif [ -e /proc/mdesk ]; then
-                de="mDesk"
-            elif [ -e /proc/neod* ]; then
-                de="NeoDesk"
-            elif [ -e /proc/zdesk* ]; then
-                de="zDesk"
-            else
-                de="Atari"
-            fi
+            for possibles in "$(ls /proc)"; do
```
Comment (dylanaraps) [inline neofetch]:
```sh
for files in /proc/*; do
```

File: neofetch [hunk 16]
```diff
@@ -1163,10 +1153,9 @@ get_cpu() {
         ;;
 
         "FreeMiNT")
-            cpu="$(cat /kern/cpuinfo | awk -F':' '/CPU:/ {printf $2}')"
-            speed="$(cat /kern/cpuinfo | awk -F':' '/Clocking:/ {printf $2}' | awk -F'.' '{ printf $1 }')"
+            cpu="$(awk -F':' '/CPU:/ {printf $2}' /kern/cpuinfo)"
```
Comment (dylanaraps) [inline neofetch]: These two `awk | /kern/cpuinfo` can probably become one statement. See the example in one of my other comments.

File: neofetch [hunk 17]
```diff
@@ -426,6 +435,12 @@ get_uptime() {
                     h="${h#0}" t="${t#0}"
                     seconds="$((d*86400 + h*3600 + ${t%%:*}*60 + ${t#*:}))"
                 ;;
+
+                "FreeMiNT")
```
Comment (dylanaraps) [inline neofetch]: Can this not be merged with the BSD/MacOS uptime detection?

Comment (dylanaraps) [reply neofetch]: I know the outputs of the same command are different but the regex matching should work on both outputs.

**Comment by ArmstrongJ** [comment]
Sorry about those file listing loops!  Your solution was far more sensible than using `ls`, which was just silly.   The `awk` calls for CPU speed and model just didn't work out as a single call for whatever reason.  I did add a fix in case a decimal in the speed information isn't present.

File: neofetch [hunk 18]
```diff
@@ -3368,6 +3434,13 @@ get_distro_colors() {
             ascii_file="freebsd"
         ;;
 
+        "FreeMiNT"*)
+            # Don't explicitly set colors since
+            # TosWin2 doesn't reset well
```
Comment (dylanaraps) [inline neofetch]: Full stop. `.`

Also, remove the commented `set_colors` line.

**Comment by dylanaraps** [comment]
LGTM :+1: 

Just fix up those comments and I'll merge. Thanks for working on this!

**Comment by ArmstrongJ** [comment]
Thanks for merging and all the help along the way!  I'm happy to be able to use neofetch on FreeMiNT now!  Much appreciated!

**Comment by dylanaraps** [comment]
No problem! Thanks for working on the PR and incorporating my feedback. :)

**Pull request merged**